bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/logic/BiweeklyPresentableCalendar.java

```diff
@@ -0,0 +1,327 @@
+                }
+            }
+            if (!cancel) {
+                return possibleCandidate;
+            }
+        }
+        return null;
+    }
+
+    private DateIterator getRecurredEventDateIterator(VEvent vEvent) {
+        TimezoneInfo tzinfo = this.usedCalendar.getTimezoneInfo();
+
+        DateStart firstStart = vEvent.getDateStart();
+        TimeZone tz;
+        if (tzinfo.isFloating(firstStart)) {
+            tz = TimeZone.getDefault();
+        } else {
+            TimezoneAssignment startAssignment = tzinfo.getTimezone(firstStart);
+            tz = (startAssignment == null ? TimeZone.getTimeZone("UTC") : startAssignment.getTimeZone());
+        }
+        return vEvent.getDateIterator(tz);
+    }
+
+    private static @Nullable Duration getEventLength(VEvent vEvent) {
+        DurationProperty duration = vEvent.getDuration();
+        if (duration != null) {
+            biweekly.util.Duration eventDuration = duration.getValue();
+            return Duration.ofMillis(eventDuration.toMillis());
+        }
+        DateStart start = vEvent.getDateStart();
+        DateEnd end = vEvent.getDateEnd();
+        if (start == null || end == null) {
+            return null;
+        }
+        return Duration.between(start.getValue().toInstant(), end.getValue().toInstant());
+    }
+
+    private static class VEventWPeriod {
+        VEvent vEvent;
+        Instant start;
+        Instant end;
+
+        VEventWPeriod(VEvent vEvent, Instant start, Instant end) {
+            this.vEvent = vEvent;
+            this.start = start;
+            this.end = end;
+        }
+
+        Event toEvent() {
+            String title;
+            Summary eventSummary = vEvent.getSummary();
+            if (eventSummary != null) {
+                title = eventSummary.getValue();
+            } else {
+                title = "-";
+            }
+            String description;
+            Description eventDescription = vEvent.getDescription();
+            if (eventDescription != null) {
+                description = eventDescription.getValue();
+            } else {
+                description = "";
+            }
+            return new Event(title, start, end, description);
+        }
+    }
+
+    /**
+     *
+     * @param frameBegin the start of the time frame
+     * @param frameEnd the start of the time frame
+     * @return list of iCalendar Events that BEGIN within the time frame
+     *
+     */
+    @Override
+    public @Nullable List<Event> getJustBegunEvents(Instant frameBegin, Instant frameEnd) {
+        List<Event> eventList = null;
+        // process all the events in the iCalendar
+        for (VEvent event : usedCalendar.getEvents()) {
+            // iterate over all begin dates
+            DateIterator begDates = getRecurredEventDateIterator(event);
+            while (begDates.hasNext()) {
+                Instant begInst = begDates.next().toInstant();
+                if (begInst.isBefore(frameBegin)) {
```

Comment:
A Duration in Java describes just an amount of time, instead of the range between two points in time. So a duration is not applicable here. Also the handling before and after an event is different, so there would be no benefit for checking if just inside or outside an time frame.